### PR TITLE
[gql][consistent-reads][6/n] Implement consistent reads for dynamic fields

### PR DESCRIPTION
**Pull request by wlmyng**
## Description 

Implement a historical read query for `DynamicField`'s `query` and `paginate` against the `objects_snapshot` and `objects_history` tables, and bound this historical read by a `checkpoint_viewed_at` for consistency. 

Substitute `DynamicField`'s internal representation with the `MoveObject` graphql type instead of the `Object` type.

This led to an observation where `MoveObject::query` returns an error on encountering a `WrappedOrDeleted` object. The underlying conversion has been modified to return `MoveObjectDowncastError::WrappedOrDeleted`, which `MoveObject::query` will convert into an `Ok(None)`.

Additionally, in order to look up the correct dynamic object fields for a parent object, we need to bound them by its parent's version. The entrypoints to fetch a dynamic object and dynamic field have been updated with an additional `parent_version: Option<u64>` - this is used as the upper bound if provided. Note that for dynamic object fields, the resolved `DynamicField` is actually the field object - so to resolve the `value` of a field object, to the value object, we must use the field object's version as the upper bound to return the correct data.

Something to call out is that when we look up a dynamic object field, the object retrieved is the `Field<Wrapper<K>, V>` which points to the expected parent object. To resolve the value of this dynamic object field, we look up the wrapped object(?) for the actual contents of the dynamic field. This move object will have its parent pointed to the wrapper, not the actual parent object.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by wlmyng** [comment]
I borked a rebase and it caused github to close the PR at https://github.com/MystenLabs/sui/pull/15934 - reopening one here.

**Comment by amnn** [comment]
FWIW, you could probably have re-opened the previous PR (but not a big deal).

File: crates/sui-graphql-rpc/src/types/owner.rs [hunk 1]
```diff
@@ -28,6 +28,9 @@ use sui_types::gas_coin::GAS;
 #[derive(Clone, Debug)]
 pub(crate) struct Owner {
     pub address: SuiAddress,
+    /// If set, this is the version of the object designated as an Owner. Used to query the dynamic
+    /// fields of the object.
+    pub version: Option<u64>,
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/owner.rs:33]: I don't think it makes sense to set this on the `Owner` field -- the idea was to add an optional `parentVersion` parameter to `dynamic_field`, `dynamic_object_field`, and `dynamic_fields`.

File: crates/sui-graphql-rpc/src/types/owner.rs [hunk 2]
```diff
@@ -36,6 +39,9 @@ pub(crate) struct Owner {
 /// Type to implement GraphQL fields that are shared by all Owners.
 pub(crate) struct OwnerImpl {
     pub address: SuiAddress,
+    /// The version of the object to use when querying dynamic fields. If not set, defaults to using
+    /// the latest checkpoint.
+    pub version: Option<u64>,
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/owner.rs:44]: Therefore you also don't need this here, you just add the version parameter to all the dynamic field functions, and pass the version up from all the sub-types.

File: crates/sui-graphql-rpc/src/types/owner.rs [hunk 3]
```diff
@@ -235,9 +241,15 @@ impl Owner {
...
             ctx.data_unchecked(),
             self.address,
-            match self.checkpoint_viewed_at {
-                Some(checkpoint_viewed_at) => ObjectLookupKey::LatestAt(checkpoint_viewed_at),
-                None => ObjectLookupKey::Latest,
+            match (self.version, self.checkpoint_viewed_at) {
+                (Some(version), checkpoint_viewed_at) => ObjectLookupKey::VersionAt {
+                    version,
+                    checkpoint_viewed_at,
+                },
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/owner.rs:248]: I mentioned in some earlier PR that I didn't understand why we needed to filter on both the version and the checkpoint we are viewing at. My understanding is that the importance of `checkpoint_viewed_at` is less to do with filtering the object, and more to do with propagating the `checkpoint_viewed_at` context. Is that right?

File: crates/sui-graphql-rpc/src/types/owner.rs [hunk 4]
```diff
@@ -437,9 +449,15 @@ impl OwnerImpl {
         name: DynamicFieldName,
     ) -> Result<Option<DynamicField>> {
         use DynamicFieldType as T;
-        DynamicField::query(ctx.data_unchecked(), self.address, name, T::DynamicField)
-            .await
-            .extend()
+        DynamicField::query(
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/owner.rs:452]: Dynamic Field and Dynamic Object Field point queries also need to be bounded by their parent version.

File: crates/sui-graphql-rpc/src/types/dynamic_field.rs [hunk 5]
```diff
@@ -266,3 +307,75 @@ pub fn extract_field_from_move_struct(
...
+        );
+        let query = query!(
+            r#"
+        SELECT candidates.*
+        FROM ({}) candidates
+        LEFT JOIN ({}) newer
+        ON (
+            candidates.object_id = newer.object_id
+            AND candidates.object_version < newer.object_version
+        )"#,
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/dynamic_field.rs:356]: nit: whitespace (indent).

File: crates/sui-graphql-rpc/src/types/dynamic_field.rs [hunk 6]
```diff
@@ -109,7 +104,10 @@ impl DynamicField {
             let obj = MoveObject::query(
                 ctx.data_unchecked(),
                 self.df_object_id,
-                ObjectLookupKey::Latest,
+                match self.super_.super_.checkpoint_viewed_at {
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/dynamic_field.rs:107]: This needs to use a version-bounded query, not a checkpoint-bounded query.

File: crates/sui-graphql-rpc/src/types/dynamic_field.rs [hunk 7]
```diff
@@ -172,66 +160,119 @@ impl DynamicField {
...
+
+        super_.map(Self::try_from).transpose()
     }
 
     /// Query the `db` for a `page` of dynamic fields attached to object with ID `parent`.
     pub(crate) async fn paginate(
         db: &Db,
         page: Page<object::Cursor>,
         parent: SuiAddress,
+        version: Option<u64>,
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/dynamic_field.rs:183]: nit: rename to `parent_version` and update the docs to explain what this is used for and why.

**Review by amnn — COMMENTED** [review]
Trending in the right direction, but I think there's still some more things that need to be addressed -- PTAL at the inline comments!  Thanks @wlmyng .

File: crates/sui-graphql-rpc/src/types/owner.rs (hunk 3 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/owner.rs:248]: The `checkpoint_viewed_at` sets an upper bound - the object at version might exist at a later checkpoint, but not at the context we are given

File: crates/sui-graphql-rpc/src/types/dynamic_field.rs (hunk 6 above)
Comment (amnn) [reply crates/sui-graphql-rpc/src/types/dynamic_field.rs:107]: This PR comment was marked as resolved, but the query still seems to be checkpoint bounded, not version bounded -- was this fixed up the stack?

File: crates/sui-graphql-rpc/src/types/dynamic_field.rs [hunk 8]
```diff
@@ -172,66 +164,126 @@ impl DynamicField {
...
-            .map_err(|e| Error::Internal(format!("Failed to derive dynamic field id: {e}")))?
-            .to_vec();
+            .map_err(|e| Error::Internal(format!("Failed to derive dynamic field id: {e}")))?;
+
+        let key = match parent_version {
+            Some(version) => ObjectLookupKey::VersionAt {
+                version,
+                checkpoint_viewed_at,
+            },
+            None => match checkpoint_viewed_at {
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/dynamic_field.rs:174]: IIRC, this is going to query for the dynamic field *at* the parent version (an equality check).  This is not correct -- it needs to query for the dynamic field with the highest version that is less than or equal to the parent version (bounds check).

**Review by amnn — APPROVED** [review]
Accepting to unblock, but PTAL at the comments around point queries -- I think both the point query of the dynamic field itself, and the point query for a dynamic object field's value are not quite correct, for slightly different reasons:

The point query for the dynamic field is using an equality check on the parent. This will not work in the following scenario:

- TX1: Create DF and add to P.
- TX2: Modify P in some other way.
- Try to read DF by name against P.

At the time of the read `P`'s version will be greater than `DF`'s -- this is why we need the bounds check (this is implemented correctly or the paginated read).

In the case of reading the value for a dynamic object field, the issue is the classic one where we only have checkpoint-level granularity so you could end up reading a later version of the object because the object was changed multiple times in the same checkpoint.

The value object is a child object of the field object, so you can again fix this by bounding the version of the value object by its parent, which, this time is the field object.

File: crates/sui-graphql-rpc/src/types/dynamic_field.rs (hunk 8 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/dynamic_field.rs:174]: Thanks @amnn  this finally clicked for me :sweat_smile:

File: crates/sui-graphql-rpc/src/types/dynamic_field.rs (hunk 6 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/dynamic_field.rs]: No, I did not understand what you were saying until now. But I think I've caught on, and I've added some tests around this

File: crates/sui-graphql-rpc/src/types/dynamic_field.rs [hunk 9]
```diff
@@ -172,66 +169,126 @@ impl DynamicField {
...
+        let key = match parent_version {
+            Some(version) => ObjectLookupKey::LatestAtParentVersion {
+                version,
+                checkpoint_viewed_at,
+            },
+            None => match checkpoint_viewed_at {
+                Some(seq) => ObjectLookupKey::LatestAt(seq),
+                None => ObjectLookupKey::Latest,
+            },
+        };
```
Comment (amnn) [inline crates/sui-graphql-rpc/src/types/dynamic_field.rs:183]: nit (optional): I'm a fan of the match on tuple pattern for this kind of thing, but this is personal preference, this looks fine to be honest:

```suggestion
        use ObjectLookupKey as K;
        let key = match (parent_version, checkpoint_viewed_at) {
            (None, None) => K::Latest,
            (None, Some(checkpoint_viewed_at)) => K::LatestAt(checkpoint_viewed_at),
            (Some(version), checkpoint_viewed_at) => K::LatestAtParentVersion {
                version,
                checkpoint_viewed_at,
            },
        }
```

**Review by amnn — APPROVED** [review]
I'm going to accept this again, because I'm so happy -- this looks great!

File: crates/sui-graphql-rpc/src/types/dynamic_field.rs (hunk 9 above)
Comment (wlmyng) [reply crates/sui-graphql-rpc/src/types/dynamic_field.rs:183]: lets do it

**Pull request merged**